lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
+ */
+DataAccessObject.patchOrCreateWithWhere =
+DataAccessObject.upsertWithWhere = function(where, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) { return connectionPromise; }
+  assert(arguments.length >= 1, 'At least one argument is required');
+  if (data === undefined && options === undefined && cb === undefined && arguments.length === 1) {
+    data = where;
+    where = {};
+  } else if (options === undefined && cb === undefined) {
+        // One of:
+        // updateAll(data, cb)
+        // updateAll(where, data) -> Promise
+    if (typeof data === 'function') {
+      cb = data;
+      data = where;
+      where = {};
+    }
+  } else if (cb === undefined) {
+        // One of:
+        // updateAll(where, data, options) -> Promise
+        // updateAll(where, data, cb)
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+  data = data || {};
+  options = options || {};
+  cb = cb || utils.createPromiseCallback();
+  assert(typeof where === 'object', 'The where argument must be an object');
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+  var hookState = {};
+  var filter = { 'where': where };
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+  var modelObject = connector._models;
+  var modelName = Object.keys(modelObject)[0];
+  var id = getIdValue(this, data);
+  var context = {
+    Model: Model,
+    query: { 'where': where },
+    hookState: hookState,
+    options: options,
+  };
+  Model.notifyObserversOf('access', context, doUpsertWithNPK);
+  function doUpsertWithNPK(err, ctx) {
+    if (err) return cb(err);
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    if (!Object.keys(data).length) {
+      process.nextTick(function() {
+        cb('data field cannot be empty');
+      });
+      return cb.promise;
+    }
+    if (connector.upsertWithWhere && !isOriginalQuery) {
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+        data = ctx.data;
+        var update = data;
+        var inst = data;
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
+        Model.applyProperties(update, inst);
+        Model = Model.lookupModel(update);
+        var connector = self.getConnector();
+        var doValidate = undefined;
+        if (options.validate === undefined) {
+          if (Model.settings.validateUpsert === undefined) {
```

Comment:
ummmm I believe `validateUpsert` was added for `upsert` for backward compatibility; we plan to remove it for LB 3.0; since `upsertWithWhere` is a new method, we don't need to have this backward compatibility. Please see our Documentation for [validateUpsert](https://docs.strongloop.com/display/public/LB/Model+definition+JSON+file)

CC: @bajtos